cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/AbstractConsumerMetaTable.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright © 2017 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.metrics.process;
+
+import co.cask.cdap.api.common.Bytes;
+import co.cask.cdap.data2.dataset2.lib.table.MetricsTable;
+import com.google.common.collect.Maps;
+
+import java.util.Map;
+import java.util.SortedMap;
+import java.util.TreeMap;
+
+/**
+ * An abstraction on persistent storage of consumer information.
+ */
+public abstract class AbstractConsumerMetaTable {
+  private static final byte[] VALUE_COLUMN = Bytes.toBytes("o");
+
+  private final MetricsTable metaTable;
+
+  public AbstractConsumerMetaTable(MetricsTable metaTable) {
+    this.metaTable = metaTable;
+  }
+
+  public synchronized <T> void save(Map<T, Long> offsets) throws Exception {
```

Comment:
this would also clean up some of the other classes below where you have to do unchecked casts